### ISSUE TITLE
Move TestUtils in a separate repo, and privatize ScalaJSClassLoader

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#3384]
Currently, our test-interface artifact exposes exactly two things that are not in `sbt-testing-interface`: `TestUtils` and `ScalaJSClassLoader`.

In 0.6.x, `ScalaJSClassLoader` was more or less required to be public, because it was part of the contract of our testing interface that the `testClassLoader` would be a `ScalaJSClassLoader`, so that one could access its `namespace` field. However, in 1.x there is no `namespace` field anymore, and `ScalaJSClassLoader` is really just a stub. Therefore, it could and should be made private.

`TestUtils`, for its part, is nothing but a helper for testing frameworks that cross-compile on JVM and JS. But if they also cross-compile to Native, they might need something more. Actually, currently Scala Native [defines `org.scalajs.testinterface.TestUtils` for itself](https://github.com/scala-native/scala-native/blob/d598130fd3975cfcaa1bafbc9580471eafe382da/test-interface/src/main/scala/org/scalajs/testinterface/TestUtils.scala), which is completely ridiculous! The functionality of `TestUtils` should probably go into a portablescala artifact.

The `TestUtils` we have in `stubs` also contains the last two `macro`s of our public API, so it would be really nice to get rid of it. (We have a couple other macros in our tests, but they don't leak outside the repo.)

Speaking of `stubs`, I believe it should go in a separate repo as well, probably also under `portablescala`.

@gzm0 @densh WDYT?

**Comment by sjrd** [comment]
This would also unblock #3356.

**Comment by densh** [comment]
@sjrd It makes perfect sense to move those to another portablescala artifact.

**Comment by gzm0** [comment]
As we have discussed about a month ago (in https://gitter.im/scala-js/scala-js/contrib), `TestUtils` does not have anything test specific and should be superseded by a pure reflection library.

I have started https://github.com/portable-scala/reflect to that effect (I have a WIP branch somewhere if you want to continue from there).

Issues with the current reflect interface I have encountered so far (on JVM)
- Checking presence of `EnableReflectiveInstantiation` is potentially slow since it needs to traverse full ancestry.
- We cannot check for Scala public anymore at runtime. So things that are scala private but JVM public are instantiatable.

**Comment by sjrd** [comment]
Ah, that's right, of course. I had forgotten about that.
Is your branch pushed somewhere online?

**Comment by gzm0** [comment]
Hmmm... Doesn't seem so. I'll push it tonight if that is OK.

**Comment by sjrd** [comment]
Sure. There's no urgency. ;)

**Comment by gzm0** [comment]
https://github.com/gzm0/reflect/tree/first-impl

**Comment by sjrd** [comment]
Thanks. I've gotten a bit further at https://github.com/portable-scala/reflect/compare/master...sjrd:first-impl

* Fixed `InvokableConstructor.newInstance` (was doing an infinite loop)
* Fixed the case of local classes, although there's still a discrepancy for a class inside a lambda (for some reason on the JVM it's not identified as a local class in that case)
* Added the shell implem for JS
* Commented out the `protected` constructors so that tests pass

It seems that constructors that are marked as `protected` or `protected[this]` when reaching the scalac back-end are anyway translated to `public` in the bytecode. We might have to adapt Scala.js' handling of `@EnableReflectiveInstantiation` to consider `protected` constructors as `public`, to allow consistency.

For the performance problem of checking whether a superclass or superinterface has `@EnableReflectiveInstantiation`, perhaps we could lift the cache to a member of `object Reflect`, so it is persisted across many calls to `lookupXYZ`. We could make it a `java.util.WeakHashMap` not to retain useless memory.

**Comment by sjrd** [comment]
After some investigation, it turns out that scalac *never* emits `protected` methods in the bytecode. See https://github.com/scala/scala/blob/82eb9f1cc1d1fa9ce2ff9a154b977734bb92c971/src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala#L715. I could trace back that code all the way to the introduction of the new byte code generator in `GenBCode`, and it was also the case in 2.11.x' old `GenICode`. So if we want to make protected constructor reflection consistent across Scala/JVM and Scala.js, we'll have to allow protected constructors in addition to public ones [here](https://github.com/scala-js/scala-js/blob/0ff1fbe3f7a003482c3b785b6a374202f44f47c1/compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala#L1035).

WDYT?

**Comment by sjrd** [comment]
Fixed by

* the separate repo https://github.com/portable-scala/portable-scala-reflect, and
* the removals from this repo in a22886bdacb80bcac749cb234fb1eac134552047.
